### ISSUE TITLE
loadbalancer algorithms

### ISSUE THREAD

**Issue by szuecs** [zalando/skipper#557]
- #536 implements a loadbalancer feature with only roundrobin algorithm. 
- After #556 we should have a solid code base to easily implement different algorithms per route. 

We want to have some other algorithms being possible and able to choose per route (for example from ingress). Algorithm ideas we could support:

- [ ] random
- [ ] LRU (adds memory usage)
- [ ] closest (by ip on the local node)
- [ ] fastest response (maybe complicated to do, because metrics lookup required)
- [ ] consistent hashing (supports sticky sessions across one TCP/IP connection)
- [ ] [Maglev](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/44824.pdf) (if possible)

**Comment by ruiaraujo** [comment]
Weighted round robin would also be a good alternative, where the weight is dependent of the size of the nodes.

**Comment by szuecs** [comment]
@ruiaraujo can you elaborate on how to pass the data you need from nodes to the load balancer?

**Comment by ruiaraujo** [comment]
Could be something like a weight on each target.

Use case
I deploy my application using Spot of multiple sizes. My apps reacts and takes advantage of the added CPUs. 

I have Skipper as LB where the route has the meta information on each node relative size and the roundRobin would take advantage of this information to generate the right amount of load for each target.

I have no idea how this would look and how it should be specified.

**Comment by szuecs** [comment]
I think a simpler algorithm that could have a similar effect is the best (least) of N choices, but we would need to have data per backend in lbcontext for the lb algorithm.

**Comment by szuecs** [comment]
@ruiaraujo I added your suggestion to the list

**Comment by ardaguclu** [comment]
I would like to work on "least requests". Because round robin is not uniformly distributed between nodes especially if long connections exist. Do you have any concern or objection before that?

**Comment by szuecs** [comment]
As discussed in slack there is https://github.com/zalando/skipper/pull/1507 that you can test or also check the code to make a new PR, if you like.

**Comment by AshrafAhmed9** [comment]
Hi @szuecs, following up on your pointer from #4096. Looking at the current state of loadbalancer/algorithm.go, several items from the list are already done: random, consistent hashing (incl. bounded loads via consistentHashBalanceFactor), least-of-N (powerOfRandomNChoices), and slow ramp-up via the fadein filter (#1207).

From what's left, weighted round robin looks like the most useful next step — @ruiaraujo's use case above (heterogeneous Spot node sizes) is still a real one.

One design question before I write any code: there's currently no channel for a static per-endpoint weight. routing.LBEndpoint only carries scheme/host/metrics, and eskip.LBEndpoint.Zone is already dropped in parseEndpoints during conversion, so weights would need new plumbing through eskip and routing, not just a new algorithm.

Where would you want the weight to come from?

- New eskip syntax on LB endpoints (e.g. <weight, "http://10.0.0.1:8080">)
- A Kubernetes annotation, similar in spirit to backend-weights but per-endpoint
- Something else you had in mind

Happy to also wire Zone through to routing.LBEndpoint while touching that conversion, if that's welcome. If the direction sounds reasonable, I'd take this one.

**Comment by MustafaSaber** [comment]
> Happy to also wire Zone through to routing.LBEndpoint while touching that conversion

This should be fixed or addressed in this PR https://github.com/zalando/skipper/pull/4111 because we need to show latency/graph by zone

I think weight should come from LB endpoints as it's a thing that we will need to update with each round of requests, I think powerOfRandomNChoices  & passive health check would give a nice idea about this and how can you update weight after each request.

**Comment by AshrafAhmed9** [comment]
Thanks @MustafaSaber, that simplifies things a lot — good to know zone is covered by #4111, I'll leave that to it.

I looked at how powerOfRandomNChoices and the passive health check consume per-endpoint state, and I see what you mean: routing.LBEndpoint.Metrics already reaches the algorithms via the EndpointRegistry, so a dynamic weight needs no eskip/Kubernetes plumbing at all. Sketch:


- add a weight to the registry entry, computed in updateStats() from the same rotating totalRequests/totalFailedRoundTrips slots the passive health check uses (success-rate based to start; latency could feed in later)
- a weightedRoundRobin (or weighted-random) algorithm in loadbalancer/algorithm.go that selects proportionally to Metrics.Weight()

One question: should this be a new algorithm name, or would you rather see it as an improvement to powerOfRandomNChoices (e.g. weighting the N-choice comparison by failure rate, not just inflight count)? Happy to go either way — @szuecs any preference before I start?

**Comment by szuecs** [comment]
I think it's a new load balancer algorithm.

**Comment by AshrafAhmed9** [comment]
Opened #4124 implementing weightedRoundRobin as discussed above.

**Comment by bedakb** [comment]
I'd like to work on implementing "least requests" if there are no any objections.

**Comment by MustafaSaber** [comment]
> I'd like to work on implementing "least requests" if there are no any objections.

Please do 😄!

**Comment by shrikant23codes** [comment]
Planning to pick up Maglev algorithm. Currently going through the paper. Let me know if anything else has to be handled here differently.

**Comment by szuecs** [comment]
@shrikant23codes it will be an interesting journey, because you need to get the 4 tuple from the tcp conn.

**Comment by shrikant23codes** [comment]
@szuecs I looked into the paper and the existing LBAlgorithm implementations. Here is what I am thinking:

We can split this into two parts. First, add the Maglev lookup table and use a temporary request key. This will test the table build, endpoint lookup, and endpoint removal behaviour.

Second, add the TCP 4-tuple key. For this, we must get source IP, source port, destination IP, and destination port from the accepted connection. If I understand correctly for a request we use this 4 tuple as the key?

This split can work if the first part keeps the key input separate from the Maglev table code.

Let me know your thoughts on this.

**Comment by szuecs** [comment]
@shrikant23codes yes makes sense to me.
I wonder how we can get the information from tcp if ServeHTTP is disconnected from this. I am also not aware of context propagation we could use for passing the data from our queuelistener to the request.
Maybe there's a hook, that we can use for example the one that we use for the new/accept/close/idle connections on the server handler or connection manager in the proxy. (I didn't check the code right now)

**Comment by shrikant23codes** [comment]
@szuecs  Will take a look at how we can get the tcp level info. 

For now will try to add maglev as a new algorithm first. Use the same key source as consistentHash: consistentHashKey if set, otherwise remote address. This would let us test the Maglev lookup table, endpoint filtering, and remapping behaviour without solving the TCP 4-tuple part in the same PR.

**Comment by szuecs** [comment]
> @szuecs  Will take a look at how we can get the tcp level info. 
> 
> For now will try to add maglev as a new algorithm first. Use the same key source as consistentHash: consistentHashKey if set, otherwise remote address. This would let us test the Maglev lookup table, endpoint filtering, and remapping behaviour without solving the TCP 4-tuple part in the same PR.

Yeah Request remote address and Listen addr (Listen address is fixed and can be passed by reading skipper.Options in skipper.go. No need for extra hoops into the tcp conn!
